sys/security/mac_do/mac_do.c

```diff
@@ -2172,59 +2179,154 @@ struct mac_do_setcred_data {
 	u_int setcred_flags;
 };
 
+struct mac_do_setuid_data {
+	struct mac_do_data_header hdr;
+	uid_t target_uid;
+};
+
+struct mac_do_setgid_data {
+	struct mac_do_data_header hdr;
+	gid_t target_gid;
+};
+
+struct mac_do_setgroups_data {
+	struct mac_do_data_header hdr;
+	int ngroups;
+	gid_t *groups;
+};
+
 static int
 mac_do_priv_grant(struct ucred *cred, int priv)
 {
-	struct mac_do_setcred_data *const data = fetch_data();
-	struct rules *rules;
-	const struct ucred *new_cred;
-	const struct rule *rule;
-	u_int setcred_flags;
-	int error;
-
 	/* Bail out fast if we aren't concerned. */
-	if (priv != PRIV_CRED_SETCRED)
-		return (EPERM);
+	switch (priv) {
```

Comment:
For all sub-blocks of this `switch`, please move the code in separate functions (e.g., `priv_grant_setcred()`, `priv_grant_user()`, `priv_grant_group()`, etc.).

This will remove indentation and separate the different recipes.  Declarations in a `switch` are valid in all the `switch` block, so you had to add a pair of braces around each block.  With this change, they won't be necessary.